### PR TITLE
update to use 2.0 of FOSHttpCache

### LINKED ISSUE #219: Update inline C
The C snippet in the [docs](http://foshttpcachebundle.readthedocs.org/en/latest/reference/configuration/headers.html) needs updating along the lines of https://gitlab.com/dbu/varnish-workshop/merge_requests/8.

### PR DESCRIPTION
**Pull request by dbu**
- Updating to FOSHttpCache 2.0
- Starting 2.0 development
- Adjust documentation for support of custom ttl in the library.

TODO
- [x] We need something to configure the http client: https://github.com/php-http/HttplugBundle/
- [x] figure out configuration for http clients in the proxy clients

fix #219

### CODE REVIEW

**Comment by ddeboer** [comment]
#214 has been merged, so we can merge this, too. I think we need to update the http-cache requirement to @dev, here.

**Comment by dbu** [comment]
right. updated - but we should first merge the bugfix #236 and then tag a last 1.3 version. this PR will start the 2.0 version of the bundle.

**Comment by dbu** [comment]
created the 1.3 branch, now getting serious :-)

lots of question still, however. mainly how to handle the http client... maybe a default service with a configured class if the user specifies neither class nor service?

File: .travis.yml [hunk 1]
```diff
@@ -36,11 +35,17 @@ matrix:
...
+      env:
+        - SYMFONY_VERSION=2.3.*
+        - COMPOSER_FLAGS="--prefer-lowest"
 
 before_script:
   - sh -c 'if [ "$SYMFONY_VERSION" != "" ]; then composer require --dev --no-update symfony/symfony=$SYMFONY_VERSION; fi;'
   - sh -c 'if [ "$FRAMEWORK_EXTRA_VERSION" != "" ]; then composer require --dev --no-update sensio/framework-extra-bundle=$FRAMEWORK_EXTRA_VERSION; fi;'
-  - composer install
+  - if [[ "$TRAVIS_PHP_VERSION" == "5.4" || "$TRAVIS_PHP_VERSION" == "hhvm" ]]; then composer remove "php-http/guzzle6-adapter" --dev --no-update; fi
+  - if [[ "$TRAVIS_PHP_VERSION" == "5.4" || "$TRAVIS_PHP_VERSION" == "hhvm" ]]; then composer require "php-http/guzzle5-adapter" --dev --no-update; fi
```
Comment (dbu) [inline .travis.yml]: as in FOSHttpCache 2.0

File: DependencyInjection/Compiler/LoggerPass.php [hunk 2]
```diff
@@ -28,10 +29,9 @@ public function process(ContainerBuilder $container)
...
         }
 
-        $subscriber = $container->getDefinition('fos_http_cache.event_listener.log')
-            ->setAbstract(false);
-
-        $container->getDefinition('fos_http_cache.cache_manager')
-            ->addMethodCall('addSubscriber', array($subscriber));
+        $container->getDefinition('fos_http_cache.event_listener.log')
+            ->setAbstract(false)
+            ->addTag('kernel.event_subscriber');
```
Comment (dbu) [inline DependencyInjection/Compiler/LoggerPass.php]: why did we not do things like this previously? and should we instead just make the logger an optional dependency and have the subscriber not subscribe to anything when logger is null?

File: DependencyInjection/FOSHttpCacheExtension.php [hunk 3]
```diff
@@ -315,11 +315,6 @@ private function loadCacheTagging(ContainerBuilder $container, XmlFileLoader $lo
         if (!empty($config['rules'])) {
             $this->loadTagRules($container, $config['rules']);
         }
-
-        $tagsHeader = $config['header'];
-        $container->getDefinition($this->getAlias().'.cache_manager')
-            ->addMethodCall('setTagsHeader', array($tagsHeader))
-        ;
```
Comment (dbu) [inline DependencyInjection/FOSHttpCacheExtension.php]: duplication for BC no longer needed

File: Resources/config/nginx.xml [hunk 4]
```diff
@@ -16,12 +16,13 @@
...
             <argument type="service" id="fos_http_cache.proxy_client.nginx.guzzle_client" on-invalid="ignore"/>
         </service>
-
+<!--
         <service id="fos_http_cache.test.client.nginx"
                  parent="fos_http_cache.test.client.abstract"
                  abstract="true">
             <argument index="0">%fos_http_cache.proxy_client.nginx.base_url%</argument>
         </service>
+-->
```
Comment (dbu) [inline Resources/config/nginx.xml]: TODO: find a solution for the http clients

File: Resources/config/test_client.xml [hunk 5]
```diff
@@ -9,15 +9,18 @@
...
+    <service id="fos_http_cache.test.client.abstract"
+             class="%fos_http_cache.test.client.class%"
+             abstract="true">
+        <argument />--><!-- base_url --><!--
             <argument type="collection">
                 <argument key="%fos_http_cache.test.client.curlopt_forbid_reuse%">true</argument>
             </argument>
         </service>
     </services>
+    -->
```
Comment (dbu) [inline Resources/config/test_client.xml]: What did we even use this client for? its called test client...

File: Resources/config/varnish.xml [hunk 6]
```diff
@@ -15,13 +15,13 @@
...
         </service>
-
+<!--
         <service id="fos_http_cache.test.client.varnish"
                  parent="fos_http_cache.test.client.abstract"
                  abstract="true">
             <argument index="0">%fos_http_cache.proxy_client.varnish.base_url%</argument>
         </service>
-
+-->
```
Comment (dbu) [inline Resources/config/varnish.xml]: TODO: find a solution for the http clients

File: Resources/doc/overview.rst [hunk 7]
```diff
@@ -8,7 +8,19 @@ This bundle is available on Packagist_. You can install it using Composer:
 
 .. code-block:: bash
 
-    $ composer require friendsofsymfony/http-cache-bundle:~1.0
+    $ composer require friendsofsymfony/http-cache-bundle:~2.0
+
+Note that the FOSHttpCache_ library needs a ``psr/http-message-implementation``
+and ``php-http/adapter-implementation``. If your project does not contain one,
+composer will complain that it did not find ``psr/http-message-implementation``.
```
Comment (dbu) [inline Resources/doc/overview.rst]: this is copy-pasted from FOSHttpCache but i think its better to duplicate than to reference at this point.

File: Test/ProxyTestCase.php [hunk 8]
```diff
@@ -72,15 +74,22 @@ protected function getHttpClient()
...
      *
-     * @param string $url     Request URL (absolute or relative)
+     * @param string $uri     Request URL (absolute or relative)
      * @param array  $headers Request HTTP headers
      * @param array  $options Request options
      *
      * @return Response
      */
-    protected function getResponse($url, array $headers = array(), $options = array())
+    protected function getResponse($uri, array $headers = array())
```
Comment (dbu) [inline Test/ProxyTestCase.php]: it looks like this is never used. the only classes extending this are ProxyTestCaseTest which only wants the client and mocks everything, and ProxyTestCaseAnnotationTest which is empty. as in, we missed to implement anything in that test.

File: Tests/Unit/CacheManagerTest.php [hunk 9]
```diff
@@ -82,6 +82,7 @@ public function testRefreshRoute()
 
     public function testTagResponse()
     {
+        $this->markTestSkipped('TODO refactor to use tag handler');
```
Comment (dbu) [inline Tests/Unit/CacheManagerTest.php]: check after https://github.com/FriendsOfSymfony/FOSHttpCache/pull/237 is sorted out

Comment (ddeboer) [inline Tests/Unit/CacheManagerTest.php]: Has been merged, so this can be updated.

**Comment by ddeboer** [comment]
As we tagged FOSHttpCache 2.0-alpha1, it would be a good time to update the bundle to work with 2.0 so users can test both in tandem. @dbu Do you have some time to resolve the merge conflicts?

> We need something to configure the http client: https://github.com/php-http/HttplugBundle/

In the library, we do most of the configuration on [HttpDispatcher](http://foshttpcache.readthedocs.io/en/2.0.0-alpha1/proxy-clients.html#basic-http-setup-with-httpdispatcher), which wraps the Httplug client; not on Httplug itself. So while I’m okay with adding HttplugBundle as a dependency for users that want to tweak the underlying Httplug client’s settings (such as timeouts etc.), I’m not sure all users need the extra bundle. For most, being able to choose your own HTTP client adapter plus the configuration options that we already have should be sufficient.

**Comment by dbu** [comment]
> As we tagged FOSHttpCache 2.0-alpha1, it would be a good time to update
> the bundle to work with 2.0 so users can test both in tandem. @dbu
> <https://github.com/dbu> Do you have some time to resolve the merge
> conflicts?

yep, agree. will try to find some time.

>     We need something to configure the http client:
>     https://github.com/php-http/HttplugBundle/
> 
> In the library, we do most of the configuration on HttpDispatcher
> <http://foshttpcache.readthedocs.io/en/2.0.0-alpha1/proxy-clients.html#basic-http-setup-with-httpdispatcher>,
> which wraps the Httplug client; not on Httplug itself. So while I’m okay
> with adding HttplugBundle as a dependency for users that want to tweak
> the underlying Httplug client’s settings (such as timeouts etc.), I’m
> not sure all users need the extra bundle. For most, being able to choose
> your own HTTP client adapter plus the configuration options that we
> already have should be sufficient.

i will have a look again at that. but i don't want to duplicate
configuration, maybe say by default we use autodiscovery otherwise
specify a http client service. and to get the service, configure
HttplugBundle or do your own thing.

File: Tests/Unit/CacheManagerTest.php [hunk 10]
```diff
@@ -88,6 +88,7 @@ public function testRefreshRoute()
      */
     public function testTagResponse()
     {
+        $this->markTestSkipped('TODO refactor to use tag handler');
```
Comment (dbu) [inline Tests/Unit/CacheManagerTest.php]: needs to be adjusted

**Comment by ddeboer** [comment]
> maybe say by default we use autodiscovery otherwise
specify a http client service. and to get the service, configure
HttplugBundle or do your own thing.

I like this approach: let’s suggest HttplugBundle (it’s good bundle :wink:) but leave it open for users to configure the service any way they like. FOSHttpCacheBundle then only needs the service id.

File: CHANGELOG.md [hunk 11]
```diff
@@ -4,11 +4,33 @@ Changelog
...
+  (http://foshttpcachebundle.readthedocs.org/en/stable/installation.html)
+
+* [Tags] The TagHandler has been split. Invalidating tags happens through the
+  CacheManager (if you use annotations for tag invalidation, you don't need to
+  change anything). Recording tags and writing them into the responses is done
+  through the SymfonyResponseTagger now.
+
+* [Test] Dropped the proxy client services as they where not used anywhere. The
+  services `fos_http_cache.test.client.varnish` and `fos_http_cache.test.client.nginx`
+  no longer exist.
```
Comment (dbu) [inline CHANGELOG.md]: @ddeboer can you agree with this? i feel that with httplug and discovery, its easy enough to do this now on your own in tests.

File: DependencyInjection/Configuration.php [hunk 12]
```diff
@@ -397,6 +314,41 @@ private function addProxyClientSection(ArrayNodeDefinition $rootNode)
...
+                    ->isRequired()
+                    ->requiresAtLeastOneElement()
+                    ->prototype('scalar')->end()
+                ->end()
+                ->scalarNode('base_url')
+                    ->defaultNull()
+                    ->info('Default host name and optional path for path based invalidation.')
+                ->end()
+                ->scalarNode('http_client')
+                    ->defaultNull()
```
Comment (dbu) [inline DependencyInjection/Configuration.php]: we don't need any reference to the httplug-bundle. by default, we pass no client to HttpDispatcher, which will prompt that to do a discovery. if discovery is not desired, a client can be configured.

**Comment by dbu** [comment]
imho this is ready to merge now. there is a bunch of things we should clean up in this bundle, but i think we should not do that in the same PR. would prefer to merge this ASAP and then do the other things in separate PRs.

File: CHANGELOG.md (hunk 11 above)
Comment (ddeboer) [inline CHANGELOG.md]: Agreed!

**Comment by ddeboer** [comment]
Ok, fixed the docs build, so almost green. There’s a dep problem on prefer-lowest, do you have an idea?

File: .travis.yml [hunk 13]
```diff
@@ -28,10 +28,11 @@ branches:
...
   include:
-    - php: 5.5
-      env:
-        - COMPOSER_FLAGS="--prefer-lowest"
-        - SYMFONY_VERSION='2.8.*'
+# fails to install. try to get this to work again before releasing
+#    - php: 5.5
+#      env:
+#        - COMPOSER_FLAGS="--prefer-lowest"
+#        - SYMFONY_VERSION='2.8.*'
```
Comment (dbu) [inline .travis.yml]: jordi thinks we hit some hard to tell edge case: https://twitter.com/seldaek/status/805698883456270336

i comment this build out for now and try again when we released the library

**Pull request merged**